Sources/FoundationNetworking/URLSession/libcurl/MultiHandle.swift

```diff
@@ -195,22 +195,19 @@ private extension URLSession._MultiHandle {
     }
 
     /// Removes socket reference from the shared store. If there is work item scheduled,
-    /// executes it on the current thread.
+    /// executes it on the current thread. Then, schedules the socket for closure.
     func endOperation(for socketReference: _SocketReference) {
         precondition(socketReferences.removeValue(forKey: socketReference.socket) != nil, "No operation associated with the socket")
         if let workItem = socketReference.workItem, !workItem.isCancelled {
             // CURL never asks for socket close without unregistering first, and
             // we should cancel pending work when unregister action is requested.
             precondition(!socketReference.shouldClose, "Socket close was scheduled, but there is some pending work left")
             workItem.perform()
-        }
-    }
 
-    /// Marks this reference to close socket on deinit. This allows us
-    /// to extend socket lifecycle by keeping the reference alive.
-    func scheduleClose(for socket: CFURLSession_socket_t) {
-        let reference = socketReferences[socket] ?? _SocketReference(socket: socket)
-        reference.shouldClose = true
+            /// Marks this reference to close socket on deinit. This allows us
+            /// to extend socket lifecycle by keeping the reference alive.
+            socketReference.shouldClose = true
+        }
```

Comment:
Should we set `.shouldClose = true` outside of the `if let workItem = ...` block? It looks like `tearDown(handle:socket:queue)` will call `handle.endOperation(for: socketReference)` where `socketReference.workItem` is `nil`.